streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -184,32 +217,21 @@ public void registerStore(final StateStore store, final StateRestoreCallback sta
 
         log.info("Restoring state for global store {}", store.name());
         final List<TopicPartition> topicPartitions = topicPartitionsForStore(store);
-        Map<TopicPartition, Long> highWatermarks = null;
 
-        int attempts = 0;
-        while (highWatermarks == null) {
-            try {
-                highWatermarks = globalConsumer.endOffsets(topicPartitions);
-            } catch (final TimeoutException retryableException) {
-                if (++attempts > retries) {
-                    log.error("Failed to get end offsets for topic partitions of global store {} after {} retry attempts. " +
-                        "You can increase the number of retries via configuration parameter `retries`.",
-                        store.name(),
-                        retries,
-                        retryableException);
-                    throw new StreamsException(String.format("Failed to get end offsets for topic partitions of global store %s after %d retry attempts. " +
-                            "You can increase the number of retries via configuration parameter `retries`.", store.name(), retries),
-                        retryableException);
-                }
-                log.debug("Failed to get end offsets for partitions {}, backing off for {} ms to retry (attempt {} of {})",
-                    topicPartitions,
-                    retryBackoffMs,
-                    attempts,
-                    retries,
-                    retryableException);
-                Utils.sleep(retryBackoffMs);
-            }
+        final Map<TopicPartition, Long> highWatermarks;
+        try {
+            highWatermarks = globalConsumer.endOffsets(topicPartitions);
+        } catch (final TimeoutException retryableException) {
+            log.debug(
+                "Failed to get end offsets for partitions {}. The broker may be transiently unavailable at the moment. Will retry.",
+                topicPartitions,
+                retryableException
+            );
+
+            // handled in `GlobalStateMangerImpl#initialize()`
+            throw retryableException;
```

Comment:
Minor note: it's confusing to track down exceptions when they are re-thrown like this, because the stacktrace would only reference L223. Even though there is a small performance penalty, it's better for maintainability to always throw a new exception like `new RetryableException(timeoutException)`.

In this particular case, it may be more appropriate just to remove the try-catch and add a `throws` declaration. There are two things that make me think this:
1. The log message here says, "will retry", but this method can have no idea whether or not it'll be retried
2. The calling method has a comment that says this method logs the error, which is also an assumption that may not survive refactoring

It seems like we can resolve all three of these maintenence problems by just moving the log message to the caller.

This feedback also applies elsewhere.